src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
+#ifdef __linux__
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+#include <goto-programs/goto_model.h>
+
+gdb_apit::gdb_apit(const char *arg)
+  : binary_name(arg), buffer_position(0), last_read_size(0)
+{
+  memset(buffer, 0, MAX_READ_SIZE_GDB_BUFFER);
+}
+
+int gdb_apit::create_gdb_process()
+{
+  if(pipe(pipe_input) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdin!");
+  }
+  if(pipe(pipe_output) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdout!");
+  }
+
+  gdb_process = fork();
+  if(gdb_process == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create gdb process.");
+  }
+  if(gdb_process == 0)
+  {
+    // child process
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+    dprintf(pipe_output[1], "binary name: %s\n", binary_name);
+    char *arg[] = {
+      const_cast<char *>("gdb"), const_cast<char *>(binary_name), NULL};
+
+    dprintf(pipe_output[1], "Loading gdb...\n");
+    execvp("gdb", arg);
+
+    // Only reachable, if execvp failed
+    int errno_value = errno;
+    dprintf(pipe_output[1], "errno in child: %s\n", strerror(errno_value));
+  }
+  else
+  {
+    // parent process
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+    write_to_gdb("set max-value-size unlimited\n");
```

Comment:
Is there a race hazard here? e.g. if you write before the child process has fully started up? Does GDB do anything to flush the pipe before it prints it's prompt? (which could potentially mean this command gets lost)

Reply:
In the latest commit we're waiting for the gdb prompt before writing the command.